### ISSUE TITLE
HomematicIP-Rest-Api v2 - Test and Feedback

### ISSUE THREAD

**Issue by hahn-th** [hahn-th/homematicip-rest-api#526]
Hello,
I am currently in the process of developing the complete library from scratch which can be accessed via Branch v2. But beware: everything is still very alpha-status :)

I want to know, how you use the library? What are your requirements? Are you interpret results from cli? Do you use the cli at all??? What are your needs?

**Example Cli:**
- Which values need to be output via the CLI? 
- Everyone could use a json file to decide for themselves what they want to output for each type.
- Does anyone use the Cli at all?

**Information about status changes:**
Many people keep running into the throttle problem when retrieving the status via the Cli. The reason for this is that the Cli is restarted again and again. This leads to the complete configuration being downloaded each time. The problem does not occur if the changes are pushed via websockets. 
I now had the following ideas:
- "Listen" mode that writes the received events to a file
- "Plugin" system with pluggy??? But i guess this would be a little too much 😆
- Connection to EventManager, which is already there

**A few things will definitely change:**
- No more dedicated classes for each Device, Channel, Group, etc. There is a base class for Device, Home, Group, FunctionalChannel and so on, but no more class for AccessControlChannel, for example. 
- The actions are "loose" and are linked to channels, etc. via Decorator. If an action is executed, the decorator is testing, if the action is applicable for the device.
- There is no "home" class anymore. Home is now just a stupid poco. There is the runner, which contains the model.
- The **Cli** is now accessible via `hmip`.
- The auth script does not exist anymore. Use `hmip auth` instead
- Just explore the Cli via `hmip --help`
- All new devices are there without the need of any code changes. But just "read-only" mode. If any rest requests should be made for the device, this must be implemented.

Any input is appreciated.

**Comment by hahn-th** [comment]
And here is the pypi package: 
pip install homematicip==1.1.0.post1.dev15

**Comment by cyraxx** [comment]
From my point of view, as the developer of the MQTT bridge (https://github.com/cyraxx/homematicip-rest-mqtt), I don't use the CLI at all, only the programmatic interface. I also haven't run into any rate limiting issues ever.

Currently my project has dedicated code for each supported device class that maps its properties to MQTT topics, and incoming topics back to device commands. With your change to a generic Device class with dynamically defined properties and actions in v2, I could possibly do away with this dedicated mapping for each device. So any new device would automatically be supported by the MQTT bridge – sounds neat!

**Comment by hahn-th** [comment]
Hi. thanks for your response. This v2 Version is obsolete. The effort to migrate HomeAssistant for v2 was to big, so i decided to stop that. Nevertheless the data of devices are stored in the functional channels. To be future proof, you could do your mapping for channels, not for devices.

by the way:
You should use a json file to configure the important attributes of each channel-type. So you dramatically improve the maintainability.

**Comment by cyraxx** [comment]
Right, I noticed after posting this that the original issue was over a year old already 😉 

Yes, good point about making the mappings more dynamic (e.g. based off JSON). The project started out with just a few heating devices for my home and then grew from there – it will make sense to switch to a dynamic mapping eventually.

**Comment by lackas** [comment]
Closing this: the v2 rewrite it was asking feedback on was stopped, as you said above in August, because migrating Home Assistant to it was too much.

The part that survived is the channel-centric direction from your reply, "the data of devices are stored in the functional channels ... do your mapping for channels, not for devices". That is how 2.x works: commands hang off the functional channels, and the Home Assistant integration has been migrating its entities to channel-based discovery. The `v2` branch stays where it is.
